### ISSUE TITLE
OPL2 research

### ISSUE THREAD

**Issue by nukeykt** [chocolate-doom/chocolate-doom#767]
Since my Nuked OPL3 emulates only OPL3 emulator, Chocolate Doom has no proper OPL2(Actually it is just OPL3 in OPL2 mode). According to MAME's ymf262.cpp there are some minor differences between OPL2 and OPL3, but i'm not sure how it affects to sound because i have no OPL2 chips. And i decided to start writing OPL2 emulator. I ordered 3xYM3812(OPL2) and 3xYM3014(DAC) chips, but they'll arrive in September. So, can someone upload some tracks played on real OPL2 chip? Tracks: D_RUNNIN, D_IN_CIT, D_E1M1, D_E2M1 played in DOS Doom; clouds.mid, town.mid played in Windows9x.

**Comment by sneakernets** [comment]
Here are the specs: http://pdf1.alldatasheet.com/datasheet-pdf/view/84281/YAMAHA/YM3812.html

From what I gather, the only real differences between the OPL2 and the OPL3 is:
* the modulator waveform on OPL2 is delayed by one
* I/O access time overall is a little longer
* Composite Sine mode was removed (deemed too difficult to use?)

Now the first one sounds like a big change, but in my experience, I've yet to encounter anyone exploiting the off-by-one modulator. Composite Sine Mode was never used by anything notable. There aren't any examples I can find online, and discussions about it on vogons and other sites boil down to "it sounds like a cool idea, but a pain in the arse to program".

**Comment by nukeykt** [comment]
Just compare AdLib and SB Pro 2(OPL3 in OPL2 mode) in this video: https://www.youtube.com/watch?v=1qp7WnTWZw4
Is the difference due to filtering?

**Comment by sneakernets** [comment]
It's very likely the case. Ad Lib card used a  RC4136,  with a simple lowpass filter (Sallen-Key maybe?). not sure of the specs on the lowpass.

**Comment by nukeykt** [comment]
Got OPL2 chip working. For me it sounds exactly like OPL3. But town.mid really sounds differently due to bug in AdLib drivers in windows. It does not set WSE bit, which alters sound on OPL2, but not on OPL3. I'll do more research when i have free time.

**Comment by nukeykt** [comment]
Here's some OPL2 samples. May be someone other will hear difference?
Doom D_E1M1:
https://www.dropbox.com/s/1incaxbvzydssbg/opl2doome1m1.wav?dl=0
Doom D_E2M1:
https://www.dropbox.com/s/i5arksflwqepxlh/opl2doome2m1.wav?dl=0
Doom 2 D_RUNNIN:
https://www.dropbox.com/s/xgyajg8d8hy0l6c/opl2doom2runnin.wav?dl=0
Doom 2 D_IN_CIT:
https://www.dropbox.com/s/uo9kat8ln1atxlr/opl2doom2in_cit.wav?dl=0

**Comment by sneakernets** [comment]
Yeah, these examples  are pretty much what I expected. if there are any differences, it'd be in the **DACs.** I do know the recommended OPL2 DAC has a little harmonic distortion, but nothing that would ever be noticed, and the dB level doesn't change it much (0.05% -0.2%). OPL3 DAC, however, does have harmonic distortion, and it gets a little worse as the dB level drops, but still less than 1% (0.25%-0.65% at -40dB) and probably not noticeable unless you have something with a very long release time. Remember, the louder the volume, the *less* the distortion, to a point.

D_E1M2 at a low volume may point these differences out. I had an FM chip in a budget Yamaha keyboard (PSR-3, I believe)  with a DAC that sounded really distorted at low volumes, which I would exploit with an amp :-)

**Comment by sneakernets** [comment]
I'm gonna do an update on this- I'm happy to say that, other than the DACs the OPL2 used, and a composite mode that DMX could never access anyway, we're pretty much good . Harekiet claims the differences are so minimal that it's not worth writing home about, but here they are.

differences between OPL2 and OPL3 not documented in Yamaha datahasheets: 
- sine table is a little different: the negative part is off by one on OPL2, fixed on OPL3. This does not seem to affect the Frequency Modulation whatsoever, it's simply too small of a change, and nothing is this accurately timed to make use of any effect that "bug" may have given anybody. 
- in order to enable selection of four different waveforms on OPL2, one must set bit 5 in register 0x01(test register). On OPL3, this bit is ignored, because CSM was removed, so 4-waveform select *always* works.  (Don't confuse this with OPL3's 8-waveform select.) 
- Envelope Generators: all 15 x rates take zero time on OPL3.  On OPL2, "15 0" and "15 1" rates take some time, albeit minuscule, while 15 2 and 15 3 rates take zero time.

- Channel calculations: Output of operator 1 is in perfect sync with output of operator 2 on OPL3; OPL2 output of operator 1 is always delayed by one sample compared to the output of operator 2. 

differences between OPL2 and OPL3 shown in the datasheets: 
- YMF262 does not support CSM mode. No one used it.

Hope that helps.
